event-sourcing/event-store-postgres/src/main/java/org/apache/james/eventsourcing/eventstore/postgres/PostgresEventStoreDAO.java

```diff
@@ -0,0 +1,122 @@
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.SNAPSHOT;
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.TABLE_NAME;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.eventsourcing.AggregateId;
+import org.apache.james.eventsourcing.Event;
+import org.apache.james.eventsourcing.EventId;
+import org.apache.james.eventsourcing.eventstore.History;
+import org.apache.james.util.ReactorUtils;
+import org.jooq.JSON;
+import org.jooq.Record;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import scala.jdk.javaapi.CollectionConverters;
+
+public class PostgresEventStoreDAO {
+    private PostgresExecutor postgresExecutor;
+    private JsonEventSerializer jsonEventSerializer;
+
+    @Inject
+    public PostgresEventStoreDAO(PostgresExecutor postgresExecutor, JsonEventSerializer jsonEventSerializer) {
+        this.postgresExecutor = postgresExecutor;
+        this.jsonEventSerializer = jsonEventSerializer;
+    }
+
+    public Mono<Void> appendAll(List<Event> events, Optional<EventId> lastSnapshot) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, AGGREGATE_ID, EVENT_ID, EVENT)
+                .valuesOfRecords(events.stream().map(event -> dslContext.newRecord(AGGREGATE_ID, EVENT_ID, EVENT)
+                    .value1(event.getAggregateId().asAggregateKey())
+                    .value2(event.eventId().serialize())
+                    .value3(convertToJooqJson(event))).collect(ImmutableList.toImmutableList()))))
+            .then(lastSnapshot.map(eventId -> insertSnapshot(events.iterator().next().getAggregateId(), eventId)).orElse(Mono.empty()));
+    }
+
+    private Mono<Void> insertSnapshot(AggregateId aggregateId, EventId snapshotId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+            .set(SNAPSHOT, snapshotId.serialize())
+            .where(AGGREGATE_ID.eq(aggregateId.asAggregateKey()))));
+    }
+
+    private JSON convertToJooqJson(Event event) {
+        try {
+            return JSON.json(jsonEventSerializer.serialize(event));
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public Mono<EventId> getSnapshot(AggregateId aggregateId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.select(SNAPSHOT)
+            .from(TABLE_NAME)
+            .where(AGGREGATE_ID.eq(aggregateId.asAggregateKey()))
+            .limit(1)))
+            .map(record -> EventId.fromSerialized(Optional.ofNullable(record.get(SNAPSHOT)).orElse(0)));
```

Comment:
How are you enforcing to take the last one?

Reply:
I read what Cassandra implementation did and then do the same with Postgres. This is from Cassandra implementation:

```
    session.prepare(QueryBuilder
      .selectFrom(EVENTS_TABLE)
      .column(SNAPSHOT)
      .whereColumn(AGGREGATE_ID).isEqualTo(bindMarker(AGGREGATE_ID))
      .build())
```
How does Cassandra implementation ensure that?

Reply:
static collumn have only one value per partition key and not one value accross the aggregate.

Reply:
is there any case in which snapshots are different while aggregate_ids are the same? From this code:

```
    session.prepare(
      update(EVENTS_TABLE)
        .setColumn(SNAPSHOT, bindMarker(SNAPSHOT))
        .whereColumn(AGGREGATE_ID).isEqualTo(bindMarker(AGGREGATE_ID))
        .build())
```
I see that snapshot are always the same for a specific aggregate id.

Reply:
Yes that is the concept of static column in cassandra

Reply:
what I mean is that thanks to the above code, snapshots of different rows are always the same if these rows have the same aggregate id without the need of "static column" feature.